### ISSUE TITLE
Readonly notebook

### ISSUE THREAD

**Issue by zvikagart** [jupyterlab/jupyterlab#3251]
The goal is to render a notebook but prevent editing.
I've tried passing `readOnly: true` to `NotebookWidgetFactory` but that doesn't seem to have any effect. I can see in the code that a readonly flag is being passed around and applied to cells, but I haven't been able to enable this feature. I also didn't find anything in the documentation.

**Comment by blink1073** [comment]
Hi @zvikagart, you should be able to set the notebook model to [readOnly](https://github.com/jupyterlab/jupyterlab/blob/master/packages/docregistry/src/registry.ts#L646).  You can do this by setting `context.model.readOnly` in the [createNew](https://github.com/jupyterlab/jupyterlab/blob/ad7b99e6bc0364d81127ab84d0be8d598bbb25c5/packages/docregistry/src/registry.ts#L910) method.

**Comment by ian-r-rose** [comment]
I suspect that @zvikagart is referring to making the codemirror instances `readOnly`. Setting the `readOnly` flag on the notebook currently seems to only affect whether JupyterLab attempts to save the file to disk: it does not prevent in-memory edits of the notebook.

I think this behavior is confusing. It can be useful in some instances, however. In the `@jupyterlab/github` extension, it allows for running of a notebook loaded from a repository without having write access to said repository. Perhaps we can make some further distinction between read-only from the server's perspective vs read-only in the UI?

**Comment by blink1073** [comment]
Ah, yep, that makes sense.

**Comment by zvikagart** [comment]
Correct, I want to prevent cells from entering edit mode, but also prevent other edits to the notebook like adding/deleting cells, moving cells around, etc.

**Comment by zvikagart** [comment]
I use the following code to hide the toolbar and set codemirror cells to `readOnly`. It prevents edits but still allows cells to go into edit mode, which is acceptable but not perfect.
```
nbWidget.toolbar.hide();
each(nbWidget.notebook.children(), (widget: any) => widget.readOnly = true);
```

I also disable execution of commands that modify the notebook. For example:
```
isReadOnly = () => false;

commands.addCommand(cmdIds.save, {
  label: 'Save',
  execute: () => nbWidget.context.save(),
  isEnabled: isReadOnly
});

commands.addCommand(cmdIds.split, {
    label: 'Split Cell',
    execute: () => NotebookActions.splitCell(nbWidget.notebook),
    isEnabled: isReadOnly
  });
```

**Comment by beledouxdenis** [comment]
I landed on this topic while I was looking for making the file editor read-only when the filesystem was mounted read-only, and I wanted to share my findings.

This is my first time digging into the source of Jupyterlab and I must say it was quite instructing.

Basically, I was able to make the file editor read-only when the file opened was not writable by adding the below line:
```
editor.setOption('readOnly', !this.context.contentsModel.writable);
```
in the method [_onContextReady](https://github.com/jupyterlab/jupyterlab/blob/2f9aed860f7d54a4dadc2b0f265ff5988e1ee1c6/packages/fileeditor/src/widget.ts#L99)  of file `packages/fileeditor/src/widget.ts`

Also, the save button remains enabled and you therefore still have the "permission denied" when you try to use it (still in the context of a read-only filesystem). I saw there is [a nicer message when the context model was flagged as `readOnly`](https://github.com/jupyterlab/jupyterlab/blob/2f9aed860f7d54a4dadc2b0f265ff5988e1ee1c6/packages/docmanager-extension/src/index.ts#L334-L337), and I therefore thought of using it as well by flagging the context model as `readOnly` in the same location:
```
contextModel.readOnly = !this.context.contentsModel.writable;
```

That being said, I am not sure you expect from a file editor to be read-only or to disable the save button when the file is read-only. Indeed, when opening a read-only file regular file editor (Gedit, Sublime Text, Atom, ...):
 - you can still edit the file, and it makes sense as you can make changes and save the file in another, writable, location (Save as),
 - you can still hit the regular `Save` button. And when you do use it, you have as well a `Permission denied` error. See below for an illustration with Atom. I guess this is done in a user-friendly context: Better let the user try to click on a button and tell him why he cannot rather than disabling the button without any explanation.
![screenshot from 2018-06-05 17-26-12](https://user-images.githubusercontent.com/5822488/40986042-928d3d24-68e5-11e8-9646-92fb5df27307.png)

My conclusion, in my opinion, is that changing these behaviors should rather be done in the context of a specific extension/add-on, for instance by overriding `_onContextReady` and adding the lines I mention above, rather than making them standard.

The error message `Permission denied` versus `Cannot Save. Document is read-only` is discussable. I myself believe `Permission denied` is more accurate: It's not because the current user cannot write on the file that another cannot.

**Comment by futurist** [comment]
@zvikagart If you want prevent editing, just override the `IDrive.get` method, to set  `metadata.editable` to `false`, like below:
```js
class ReadonlyDrive extends Drive implements Contents.IDrive {
    name = 'readonly';
    async get(
      path: string,
      options?: Contents.IFetchOptions
    ): Promise<Contents.IModel> {
        // ... get file content
        content.cells.forEach((cell: any) => {
            cell.metadata = cell.metadata || {};
            cell.metadata.editable = false;
        });
        return {
            name, path, content, ...
        } as Contents.IModel;
    }
}
```

This way everything will be readOnly.

**Comment by futurist** [comment]
I think we need a option in `SessionManager` or `ServiceManager`, if the whole notebook is readOnly and just for preview, **do not need to connect to a kernel**, and no `listRunning`.

**Comment by krassowski** [comment]
see also:
- https://github.com/jupyterlab/jupyterlab/issues/12960
- https://github.com/jupyterlab/jupyterlab/issues/18369
- https://github.com/jupyterlab/jupyterlab/issues/18677
- https://github.com/jupyterlab/jupyterlab/issues/9073

**Comment by krassowski** [comment]
https://github.com/jupyterlab/jupyterlab/pull/18680 fixes the problem of `cell.readOnly = true` having no effect in `createCodeCell()`, `createMarkdownCell()` or `createRawCell()` methods of `NotebookPanel.IContentFactory` due to model sync overriding them.

I believe that the feature proposed here should be called **view-only** mode, to distinguish it from already implemented **read-only** mode in which you can modify content but cannot save except for "save as" option; this is standard nomenclature is believe.

In order to have a proper view-only mode we would also need to:
- prevent kernel from starting up (as mentioned above); this is now possible by setting `kernelPreference: null` when opening the notebook
- hide or disable toolbar items which are irrelevant
